File: practices/r2dbc-webflux-integration-test/src/main/java/com/grizz/wooman/webflux/controller/UserController.java
package com.grizz.wooman.webflux.controller;

import com.grizz.wooman.webflux.common.User;
import com.grizz.wooman.webflux.controller.dto.ProfileImageResponse;
import com.grizz.wooman.webflux.controller.dto.SignupUserRequest;
import com.grizz.wooman.webflux.controller.dto.UserResponse;
import com.grizz.wooman.webflux.service.UserService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Mono;

@Slf4j
@RequiredArgsConstructor
@RequestMapping("/api/users")
@RestController
public class UserController {
    private final UserService userService;

    @GetMapping("/{userId}")
    public Mono<UserResponse> getUserById(
            @PathVariable String userId
    ) {
        return ReactiveSecurityContextHolder
                .getContext()
                .flatMap(context -> {
                    String name = context.getAuthentication().getName();

                    if (!name.equals(userId)) {
                        return Mono.error(new ResponseStatusException(HttpStatus.UNAUTHORIZED));
                    }

                    return userService.findById(userId)
                            .map(this::map)
                            .switchIfEmpty(
                                    Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND))
                            );
                });
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    public Mono<UserResponse> signupUser(
            @RequestBody SignupUserRequest request
    ) {
        return userService.createUser(
                        request.getName(),
                        request.getAge(),
                        request.getPassword(),
                        request.getProfileImageId())
                .map(this::map)
                .onErrorMap(RuntimeException.class, e -> {
                    return new ResponseStatusException(HttpStatus.BAD_REQUEST);
                })
                .switchIfEmpty(
                        Mono.error(new ResponseStatusException(HttpStatus.BAD_REQUEST))
                );
    }

    private UserResponse map(User user) {
        return new UserResponse(
                user.getId(),
                user.getName(),
                user.getAge(),
                user.getFollowCount(),
                user.getProfileImage().map(image ->
                        new ProfileImageResponse(
                                image.getId(),
                                image.getName(),
                                image.getUrl()))
        );
    }
}

File: practices/coroutine-r2dbc-webflux/src/main/kotlin/com/grizz/wooman/coroutine/controller/UserController.kt
package com.grizz.wooman.coroutine.controller

import com.grizz.wooman.coroutine.service.UserCoroutineService
import com.grizz.wooman.webflux.common.Image
import com.grizz.wooman.webflux.common.User
import com.grizz.wooman.webflux.controller.dto.ProfileImageResponse
import com.grizz.wooman.webflux.controller.dto.SignupUserRequest
import com.grizz.wooman.webflux.controller.dto.UserResponse
import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.core.context.SecurityContext
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Mono

@RequestMapping("/api/users")
@RestController
class UserController(
    private val userService: UserCoroutineService,
) {
    @GetMapping("/{userId}")
    suspend fun getUserById(
        @PathVariable userId: String
    ): UserResponse {
        val context = ReactiveSecurityContextHolder
            .getContext()
            .awaitSingle()

        val name = context.authentication.name
        if (name != userId) {
            throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
        }

        return userService.findById(userId)
            ?.let(this::map)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    suspend fun signupUser(
        @RequestBody request: SignupUserRequest
    ): UserResponse {
        return userService.createUser(
            request.name, request.age,
            request.password, request.profileImageId
        ).let(this::map)
    }

    private fun map(user: User): UserResponse {
        return UserResponse(
            user.id,
            user.name,
            user.age,
            user.followCount,
            user.profileImage.map { image: Image ->
                ProfileImageResponse(
                    image.id,
                    image.name,
                    image.url
                )
            }
        )
    }
}